File: chapter01/checkPermutation/checkPerm.java
import java.util.*;

public class checkPerm {

    public static boolean checkPerm(String s1, String s2) {
        if (s1.length() != s2.length()) {
            return false;
        }

        int[] charCount = new int[128];

        for (int i = 0; i < s1.length(); i++) {
            charCount[s1.charAt(i)]++;
            charCount[s2.charAt(i)]--;
        }

        for (int count : charCount) {
            if (count != 0) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        String s1 = "abc";
        String s2 = "cba";
        System.out.println(checkPerm(s1, s2));

        String s3 = "abc";
        String s4 = "def";
        System.out.println(checkPerm(s3, s4));
    }
}

File: chapter01/checkPermutation/checkPerm.ts
function checkPerm(s1: string, s2: string): boolean {
  if (s1.length !== s2.length) {
    return false;
  }

  return permutation(s1, s2);
}

function permutation(s1: string, s2: string): boolean {
  return s1.split('').sort().join('') === s2.split('').sort().join('');
}

console.log(checkPerm("aab", "aba")); //true
console.log(checkPerm("abc", "cba")); // true
console.log(checkPerm("abc", "qwe")); // false
